File: scripts/define_primer.R
# Script to indentify the primer sequences depending on the length of th primers (offset) or the sequence of the primers.
primer  <- read.table(snakemake@input[['primer_table']], header=T, stringsAsFactors=F, sep= ',')
comb.f = c() ## forward primer
comb.r = c() ## reverse primer
## pandaseq just takes the length of the primers but not the sequence, in case primer are different
## and can' be expressed by common 1L code
if(snakemake@config[['qc']][['primer_offset']] == TRUE){
  if (snakemake@config[['qc']][['mid_rm']]){ ## only forward primer
    for(i in 1:length(primer[,1])){
      tmp = primer[i,"specific_forward_primer"]
      tmp = unlist(strsplit(tmp, split = ""))
      comb.f <- c(comb.f, length(tmp))
      tmp = paste0(primer[i, "poly_N_rev"], primer[i, "specific_reverse_primer"], collapse = "")
      tmp = unlist(strsplit(tmp, split = ""))
      comb.r <- c(comb.r, length(tmp))
    }
  }else{ ### takes primer sequence
    for(i in 1:length(primer[,1])){
      tmp = paste0(primer[i,"poly_N"], primer[i,"MID"],primer[i,"specific_forward_primer"], collapse = "")
      tmp = unlist(strsplit(tmp, split = ""))
      comb.f <- c(comb.f, length(tmp))
      tmp = paste0(primer[i, "poly_N_rev"], primer[i, "specific_reverse_primer"], collapse = "")
      tmp = unlist(strsplit(tmp, split = ""))
      comb.r <- c(comb.r, length(tmp))
    }
  }
  primer <- cbind(primer, f_primer = comb.f, r_primer = comb.r)
}else{ # work with sequences not offset
  if (snakemake@config[['qc']][['mid_rm']]){
    for(i in 1:length(primer[,1])){
      comb.f <- c(comb.f,  as.character(primer[i,"specific_forward_primer"]))
      comb.r <- c(comb.r, as.character(primer[i, "specific_reverse_primer"]))
    }
  }else{
    for(i in 1:length(primer[,1])){
      comb.f <- c(comb.f, paste( primer[i,"MID"], primer[i,"specific_forward_primer"], sep=""))
      comb.r <- c(comb.r, as.character(primer[i, "specific_reverse_primer"]))
    }
  }
  primer <- cbind(primer, f_primer = comb.f, r_primer = comb.r)
  write.csv(primer, file = snakemake@output[[1]], row.names = FALSE)
}

File: scripts/define_primer.py
import pandas as pd

# Script to define the primer used by pandaseq. The defined
# primer depends on the existence of the barcode, if the reads are
# single or paired end and if pandaseq should use an offset of
# the primer length instead of the sequence. Using an offset can
# be useful if the sequence has many uncalled bases in the primer
# region, preventing a nucleotide primer from matching.

primertable = pd.read_csv(str(snakemake.input), index_col="Probe")

if snakemake.params.all_removed:
    pass
else:
    if snakemake.params.paired_end:
        if snakemake.params.offset:
            if snakemake.params.bar_removed:
                primertable["f_primer"] = (
                    primertable[["poly_N",
                    "specific_forward_primer"]].sum(axis=1).str.len())
                primertable["r_primer"] = (
                    primertable[["poly_N_rev",
                    "specific_reverse_primer"]].sum(axis=1).str.len())
            else:
                primertable["f_primer"] = (
                    primertable[["poly_N", "Barcode_forward",
                    "specific_forward_primer"]].sum(axis=1).str.len())
                primertable["r_primer"] = (
                    primertable[["poly_N_rev","Barcode_reverse",
                    "specific_reverse_primer"]].sum(axis=1).str.len())
        else:
            if snakemake.params.bar_removed:
                primertable["f_primer"] = (
                    primertable["specific_forward_primer"])
                primertable["r_primer"] = (
                    primertable["specific_reverse_primer"])
            else:
                primertable["f_primer"] = (
                    primertable[["Barcode_forward",
                    "specific_forward_primer"]].sum(axis=1))
                primertable["r_primer"] = (
                    primertable[["Barcode_reverse",
                    "specific_reverse_primer"]].sum(axis=1))
    else:
        if snakemake.params.offset:
            if snakemake.params.bar_removed:
                primertable["f_primer"] = (
                    primertable[["poly_N",
                    "specific_forward_primer"]].sum(axis=1).str.len())
            else:
                primertable["f_primer"] = (
                    primertable[["poly_N", "Barcode_forward",
                    "specific_forward_primer"]].sum(axis=1).str.len())
        else:
            if snakemake.params.bar_removed:
                primertable["f_primer"] = (
                    primertable["specific_forward_primer"])
            else:
                primertable["f_primer"] = (
                    primertable[["Barcode_forward",
                    "specific_forward_primer"]].sum(axis=1))

primertable.to_csv(str(snakemake.output))
